### PR TITLE
#18701: SQL compatibility: provide POSITION(needle IN haystack) syntax

### LINKED ISSUE #18701: SQL compatibility: provide POSITION(needle IN haystack) syntax.
**Use case**
`SELECT POSITION('/' IN s) FROM (SELECT 'Hello/World' AS s);`

The same as `position(s, '/')`

### PR DESCRIPTION
**Pull request by zhangjmruc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Function `position` now supports `position(needle in haystack)` synax for SQL compatibility. This closes #18701.
...

Detailed description / Documentation draft:
Enhanced function `position` to support SQL compatibility syntax: position(needle in haystack), which is same as position(haystack, needle).
...

closes #18701

### CODE REVIEW

File: src/Parsers/ExpressionElementParsers.cpp [hunk 1]
```diff
@@ -343,6 +343,22 @@ bool ParserFunction::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
...
+    {
+        /// POSITION(needle IN haystack) is equivalent to function position(haystack, needle)
+        const auto & list = expr_list_args->as<ASTExpressionList &>();
+        if (list.children.size() == 1)
+        {
+            const auto & in_func = list.children[0]->as<ASTFunction &>();
+            if (in_func.name == "in")
+            {
+                // switch the two arguments
+                const auto & arg_list = in_func.arguments->as<ASTExpressionList &>();
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:356]: Need a check for the number of arguments.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 2]
```diff
@@ -343,6 +343,22 @@ bool ParserFunction::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
         throw Exception("Argument of function toDate is unquoted: toDate(" + contents_str + "), must be: toDate('" + contents_str + "')"
             , ErrorCodes::SYNTAX_ERROR);
     }
+    else if (Poco::toLower(getIdentifierName(identifier)) == "position")
+    {
+        /// POSITION(needle IN haystack) is equivalent to function position(haystack, needle)
+        const auto & list = expr_list_args->as<ASTExpressionList &>();
+        if (list.children.size() == 1)
+        {
+            const auto & in_func = list.children[0]->as<ASTFunction &>();
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:352]: Need to check that it's actually a function.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 3]
```diff
@@ -343,6 +343,22 @@ bool ParserFunction::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
         throw Exception("Argument of function toDate is unquoted: toDate(" + contents_str + "), must be: toDate('" + contents_str + "')"
             , ErrorCodes::SYNTAX_ERROR);
     }
+    else if (Poco::toLower(getIdentifierName(identifier)) == "position")
+    {
+        /// POSITION(needle IN haystack) is equivalent to function position(haystack, needle)
+        const auto & list = expr_list_args->as<ASTExpressionList &>();
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:349]: Need to check that it's actually an expression list.

File: tests/queries/0_stateless/00233_position_function_family.sql [hunk 4]
```diff
@@ -13,6 +13,14 @@ select 1 = position('a', '', 1);
 select 2 = position('a', '', 2);
 select 0 = position('a', '', 3);
 
+select 1 = position('' in '');
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/00233_position_function_family.sql:16]: It's better to add a new file with test cases than editing an existing test.

**Comment by alexey-milovidov** [comment]
Could you please also rebase with upstream master (because there are some "extra" commits).

**Comment by zhangjmruc** [comment]
There is still one "extra" commit. Would you please help me on how to fix it?

**Comment by alexey-milovidov** [comment]
Ok, I will deal with it when I will merge.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Thank you! Now it is perfect.

**Comment by AnaUvarova** [comment]
Hi, @zhangjmruc !
Could you please help me understand how exactly this new functionality works?
Maybe some interesting details about it?

**Comment by zhangjmruc** [comment]
@AnaUvarova This new function POSITION(needle IN haystack) works exactly same as POSITION(haystack, needle). I don't dig in how it works, what I do is related to only some parser works to allow parser to accept the new syntax. 
> Could you please help me understand how exactly this new functionality works?
> Maybe some interesting details about it?

In parser phase, the argument in the new function needle IN haystack is parsed as an IN function, so I only need to switch the two arguments of the IN func and put them in the argument list of POSITION. The remaining handle in Interprepreter and execution phase is same as the old function position(haystack, needle). I didn't dig into.

**Pull request merged**